### PR TITLE
Added support for third-party plugins for MLFlow server

### PR DESCRIPTION
**Pull request by abaveja313**
## What changes are proposed in this pull request?
- Register third party endpoints in mlflow/server/handlers.py

(Please fill in changes proposed in this fix)

### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

Third party Plugins can now be used in the `mlflow server` command

### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by smurching** [comment]
@abaveja313 thanks for the PR! How did you test out this change / were you able to test it with a custom plugin? What use case are you trying to satisfy with the PR?

**Comment by abaveja313** [comment]
Hi @smurching ,
Thanks for your response!

After @mateiz reccomended that I create a plugin for Splice Machine (I wanted to get Splice Machine SQLAlchemy support for MLFlow Server), I realized that there isn't any support for MLFlow Server for external plugins. Thus, I made this PR to provide a fix. I have tested this with our splice machine plugin, and it works fine. would you like me to add test cases for this?

**Comment by smurching** [comment]
@abaveja313 gotcha thanks for running the manual test. It would be nice to have some level of test coverage for this - for now I've pushed a commit adding a unit test that verifies that the MLflow server recognizes custom plugins when constructing its backing store. If you could take a look a review pass on the test that'd be great, otherwise the PR LGTM :)

**Comment by smurching** [comment]
Actually, looks like I can't directly add the test, but I opened a PR against your branch adding one: https://github.com/abaveja313/mlflow/pull/1/files. Thanks!

**Comment by abaveja313** [comment]
Test looks great! Thanks so much @smurching . When you are ready, we could move forward to a merge?

**Comment by smurching** [comment]
Yep sounds good! Will merge once tests pass

**Comment by abaveja313** [comment]
Awesome!

**Comment by abaveja313** [comment]
Hi @smurching -- I was able to run the test locally on my laptop, and it passed. Looked closer at the travis logs, and I couldn't seem to figure out why it was failing (especially since the environment does not let you test python to try to debug the error). It seems like the mlflow-test-plugin was installed-- my first instinct is that maybe the environment variable is not resolving. As someone who has a vast familiarity in this codebase, do you have any idea why this test case might be failing? I will keep looking and post another comment if I find anything.

**Comment by abaveja313** [comment]
Hey @smurching -- I was able to run your test case locally, and it passed. Unfortunately, I was unable to reproduce the error observed in Travis CI. It seems as if the the travis job is not using the store specified in the plugin (even though it can obviously find it, or it would've thrown an MLFlowException, saying that the scheme is unknown). Instead, it is retrieving a SQLAlchemy store for some reason, even though the the scheme (file-plugin) does not match the beginnings of any of the supported database types. You have been so helpful throughout this PR process, and I am wondering if you could provide me any insight into what might be happening/point me in the right direction? Sorry, thanks!

File: tests/server/test_handlers.py [hunk 1]
```diff
@@ -108,3 +111,17 @@ def test_handler():
...
+@pytest.mark.large
+def test_mlflow_server_with_installed_plugin(tmpdir):
+    """This test requires the package in tests/resources/mlflow-test-plugin to be installed"""
+    from mlflow_test_plugin import PluginFileStore
+
+    env = {
+        BACKEND_STORE_URI_ENV_VAR: "file-plugin:%s" % tmpdir.strpath,
+    }
+    with mock.patch.dict(os.environ, env):
+        plugin_file_store = mlflow.server.handlers._get_store()
```
Comment (smurching) [inline tests/server/test_handlers.py]:
```suggestion
        mlflow.server.handlers._store = None
        plugin_file_store = mlflow.server.handlers._get_store()
```

Comment (smurching) [inline tests/server/test_handlers.py]:
```suggestion
        mlflow.server.handlers._store = None
        try:
            plugin_file_store = mlflow.server.handlers._get_store()
        finally:
            mlflow.server.handlers._store = None
```

**Comment by smurching** [comment]
@abaveja313 ah yea that's annoying, seems to be because we cache the store in a global variable in `handlers.py`, and this global is being set by other tests before `test_handlers.py` runs. I suspect clearing that global (as in my suggestion^) will fix the test

**Comment by smurching** [comment]
@abaveja313 thanks for the quick update! LGTM, merging

**Pull request merged**